### PR TITLE
Use the MongoDB Java driver's DNS SPI to enable mongo+srv:// connection strings in native mode 

### LINKED ISSUE #26387: Use the MongoDB Java driver's DNS SPI to enable mongo+srv:// connection strings in native mode
### Description

This feature enables full DNS lookup support in native mode without the need for workarounds and substitutions.

### Implementation ideas

This change replaces the substitutions with SPI API.

### PR DESCRIPTION
**Pull request by evanchooly**
fixes #26387

### CODE REVIEW

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I'm not an expert on DNS so if @cescoffier can have a look at the DNS part it would be great.

I added some suggestion on code simplification but overall it looks good to me.

I wonder if the config related code could be simplified by moving all the default options as default inside MongodbConfig and recovering the `MongodbConfig` bean from Arc. This would allow to directly use the config attribute instead of using the `Config` object.

File: integration-tests/mongodb-client/pom.xml [hunk 1]
```diff
@@ -120,5 +114,31 @@
...
     </build>
 
+    <profiles>
+        <profile>
+            <id>test-mongodb</id>
+            <activation>
+                <property>
+                    <name>test-containers</name>
+                </property>
+            </activation>
```
[the remark this answers is not in the document]
Comment (evanchooly) [reply integration-tests/mongodb-client/pom.xml:124]: it was suggested by clement to get the integration tests properly running.  but i think you're right.  i'll pull it and make sure it all works.  I think the actual fix was elsewhere.

File: integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/ReactiveBookResource.java [hunk 2]
```diff
@@ -22,6 +25,15 @@ private ReactiveMongoCollection<Book> getCollection() {
         return client.getDatabase("books").getCollection("my-reactive-collection", Book.class);
     }
 
+    @DELETE
+    public Response clearCollection() {
```
[the remark this answers is not in the document]
Comment (evanchooly) [reply integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/ReactiveBookResource.java:29]: good catch.  done.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                    .orElse(53);
+            dnsClientOptions
+                    .setHost(server)
+                    .setPort(port);
+        }
+        dnsClient = vertx.createDnsClient(dnsClientOptions);
+    }
+
+    private static List<String> nameServers() {
+        String file = "/etc/resolv.conf";
```
[the remark this answers is not in the document]
Comment (evanchooly) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java:76]: It absolutely won't but that's how the substitution code did it.  this is just that code relocated.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        if (!new File(file).isFile()) {
```
[the remark this answers is not in the document]
Comment (evanchooly) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java:77]: again, this is just relocated code but i'm happy to clean it up a bit if it doesn't complicate the diff for @cescoffier  to look over.

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 5]
```diff
@@ -99,6 +98,22 @@ AdditionalIndexedClassesBuildItem includeBsonTypesToIndex() {
...
+        return new AdditionalIndexedClassesBuildItem(
+                MongoDnsClientProvider.class.getName(),
+                MongoDnsClient.class.getName());
+    }
+
+    @BuildStep
+    public void registerDnsProvider(BuildProducer<GeneratedResourceBuildItem> buildProducer,
+            BuildProducer<NativeImageResourceBuildItem> nativeProducer) {
+        buildProducer.produce(
+                new GeneratedResourceBuildItem("META-INF/services/" + DnsClientProvider.class.getName(),
```
Comment (cescoffier) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:112]: Why do we need this to be generated?

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java [hunk 6]
```diff
@@ -52,7 +60,10 @@ public class MongodbConfig {
      *
      * <strong>IMPORTANT:</strong> The resolution may be different in JVM mode using the default (JNDI-based) DNS resolver,
      * and in native mode. This feature is experimental.
+     *
+     * @deprecated
```
Comment (cescoffier) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java:64]: Maybe explain why.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (cescoffier) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java:77]: No problem for me, it should be simplified.

**Review by loicmathieu — APPROVED** [review]
@evanchooly as the code is a copy from the substitution, I approve the PR so if you don't have time to look at my other comments feel free to merge it.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java [hunk 7]
```diff
@@ -14,6 +14,14 @@ public class MongodbConfig {
 
     public static final String CONFIG_NAME = "mongodb";
 
+    public static final String NATIVE_DNS_LOG_ACTIVITY = "native.dns.log-activity";
```
Comment (cescoffier) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java:17]: Isn't the new resolver used also in JVM mode?

**Review by cescoffier — DISMISSED** [review]
If I understand correctly, this PR switches to the Vert.x resolver in JVM and native mode.
I think we should provide an option where we allow using the default (JNDI-based) one in JVM mode, at least for compatibility purpose.

**Comment by evanchooly** [comment]
> > If I understand correctly, this PR switches to the Vert.x resolver in JVM and native mode. I think we should provide an option where we allow using the default (JNDI-based) one in JVM mode, at least for compatibility purpose.
> 
> I was wondering about this but since it's SPI driven that gets rather complicated. I suppose we could switch inside the `DnsProvider` and that's reasonably easy to do. Personally, I kind of like the consistency of using it exclusively.

What if we do a class availability check on the JNDI types, use the driver's stuff if JNDI is present, or this "new" stuff if it's not?  Then in JVM mode, it'd just use the driver's stuff and everything's peachy.  but in native mode, JNDI wouldn't be there so we'd use this?

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            return Collections.emptyList();
+        }
+        Path p = Paths.get(file);
+        List<String> nameServers = new ArrayList<>();
+        String line;
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(Files.newInputStream(p)))) {
+            while ((line = br.readLine()) != null) {
```
[the remark this answers is not in the document]
Comment (evanchooly) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/dns/MongoDnsClient.java]: I cleaned it up but I may have gone too far some tastes. I, obviously, prefer this approach but I can retool it if there are objections.  I'll do a squash commit locally and push up the (hopefully) final version.)

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java (hunk 7 above)
Comment (cescoffier) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java:17]: You can introduce a new property and alias that one. @radcortez explained to me that a long time ago.

Comment (radcortez) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongodbConfig.java:17]: You can use a Relocate: https://smallrye.io/smallrye-config/2.10.2-SNAPSHOT/extensions/relocate/

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java (hunk 5 above)
Comment (cescoffier) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:112]: That's probably enough to hard code it.

**Comment by evanchooly** [comment]
i took a detour on other things but this is ready, i think.

**Comment by evanchooly** [comment]
@cescoffier what do you think of this now?

**Pull request merged**